services/text_extractor_service.py

```diff
@@ -11,21 +11,30 @@
 
 
 class TextExtractorService:
-    def __init__(self, db, executor):
+    def __init__(self, db, executor, redis):
         self.executor = executor
         self.stored_recipes_service = StoredRecipesService(db)
+        self.redis = redis
 
     def find_recipe_in_url(self, form):
         url = form['url']
         instructions = form['instructions'].lower() == "true"
         ingredients = form['ingredients'].lower() == "true"
         try:
+            if self.redis.hexists(name="frequently_used_recipes", key=url):
+                return jsonify(json.loads(self.redis.hget(name="frequently_used_recipes", key=url)))
             if self.stored_recipes_service.url_in_db(url):
-                return jsonify(self.stored_recipes_service.get_recipe_from_db(url, ingredients, instructions))
-            all_text, title = text_extractor.get_all_text_from_url(url=url)
+                recipe = self.stored_recipes_service.get_recipe_from_db(url, ingredients, instructions)
+                self.cache_recipe(recipe, url)
+                return jsonify(recipe)
+
```

Comment:
sorry, you still "coupled" the redis code with text extractor service.
think if in few months we gonna switch to a new one and we DO NOT want to change text_extractor_service at all (we just want to change one dao for another.

Reply:
I think I got it. made the caching mechanism much more generic now.